File: edukate-backend/src/main/java/io/github/sanyavertolet/edukate/backend/entities/Submission.java
package io.github.sanyavertolet.edukate.backend.entities;

import io.github.sanyavertolet.edukate.common.SubmissionStatus;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

@Data
@NoArgsConstructor
@Document(value = "submissions")
public class Submission {
    @Id
    private String id;
    private String problemId;
    private String userId;
    private SubmissionStatus status;
    private List<String> fileObjectIds;

    @CreatedDate
    private Instant createdAt;

    public Submission(String problemId, String userId) {
        this.problemId = problemId;
        this.userId = userId;
        this.status = SubmissionStatus.PENDING;
        this.fileObjectIds = new ArrayList<>();
    }

    public static Submission of(String problemId, String userId) {
        return new Submission(problemId, userId);
    }
}

File: edukate-backend/src/main/kotlin/io/github/sanyavertolet/edukate/backend/entities/Submission.kt
package io.github.sanyavertolet.edukate.backend.entities

import io.github.sanyavertolet.edukate.common.ContentLanguage
import io.github.sanyavertolet.edukate.common.SubmissionStatus
import java.time.Instant
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.Id
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.relational.core.mapping.Table

@Table("submissions")
data class Submission(
    @Id val id: Long? = null,
    val problemId: Long,
    val userId: Long,
    val status: SubmissionStatus,
    val language: ContentLanguage = ContentLanguage.RU,
    val fileObjectIds: List<String> = emptyList(),
    @CreatedDate val createdAt: Instant? = null,
    @LastModifiedDate val updatedAt: Instant? = null,
) {
    fun withFileObjectIds(fileObjectIds: List<String>): Submission = copy(fileObjectIds = fileObjectIds.toList())

    companion object {
        @JvmStatic
        fun of(problemId: Long, userId: Long, language: ContentLanguage = ContentLanguage.RU) =
            Submission(problemId = problemId, userId = userId, status = SubmissionStatus.PENDING, language = language)
    }
}
